// astro/dist/core/build/plugins/plugin-middleware.js
import { MIDDLEWARE_PATH_SEGMENT_NAME } from "../../constants.js";
import { addRollupInput } from "../add-rollup-input.js";
const MIDDLEWARE_MODULE_ID = "@astro-middleware";
const EMPTY_MIDDLEWARE = "\0empty-middleware";
function vitePluginMiddleware(opts) {
  return {
    name: "@astro/plugin-middleware",
    options(options) {
      return addRollupInput(options, [MIDDLEWARE_MODULE_ID]);
    },
    async resolveId(id) {
      if (id === MIDDLEWARE_MODULE_ID) {
        const middlewareId = await this.resolve(
          `${opts.settings.config.srcDir.pathname}/${MIDDLEWARE_PATH_SEGMENT_NAME}`
        );
        if (middlewareId) {
          return middlewareId.id;
        } else {
          return EMPTY_MIDDLEWARE;
        }
      }
      if (id === EMPTY_MIDDLEWARE) {
        return EMPTY_MIDDLEWARE;
      }
    },
    load(id) {
      if (id === EMPTY_MIDDLEWARE) {
        return "export const onRequest = undefined";
      }
    }
  };
}
function pluginMiddleware(opts) {
  return {
    build: "ssr",
    hooks: {
      "build:before": () => {
        return {
          vitePlugin: vitePluginMiddleware(opts)
        };
      }
    }
  };
}
export {
  MIDDLEWARE_MODULE_ID,
  pluginMiddleware,
  vitePluginMiddleware
};
